proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,447 @@
+
+### Transfering closed accounts to the system program
+
+While it would be more intuitive to give the program account back to the system
+program immediately in the close instruction, this would also allow a program
+to reopen itself under a different loader in the same transaction, while it is
+still running. In order to prevent this the program account is only cleared and
+the rent collection mechanism will then implicitly give the account back to the
+system program at the very end of the transaction.
+
+### Removing the redeployment cooldown
+
+The cooldown enforces that each pair of an address and a slot can uniquely
+identify a version of a program, which simplifies caching logic. Thus this
+one slot cooldown will be kept in loader-v4.
+
+### Inferring the program length from the ELF header
+
+There are two issues with that: First, a ELF header does not know how long
+its file is, it has to be parsed and calculated from offsets of various
+sections, which increases the validator implementation complexity. Second,
+this would require the ELF header to always be uploaded first and always in a
+single chunk, wich increases the complexity of the uploading logic. Instead we
+choose the user having to announce the length of the program to be uploaded
+explicitly before the actual upload starts.
+
+## New Terminology
+
+The _current slot_ is as in the Clock sysvar.
+
+## Detailed Design
+
+The feature gate must enable loader-v4 program management and execution.
+
+### Owned Program Accounts
+
+Accounts of programs owned by loader-v4 must have the following layout:
+
+- Header (which is 48 bytes long):
+  - `u64` status enum:
+    - Enum variant `0u64`: `Invalid`, account was zero-filled externally
+    - Enum variant `1u64`: `NeverBeenDeployed`, used as write buffer
+    - Enum variant `2u64`: `Retracted`, program is in maintenance
+    - Enum variant `3u64`: `Deployed`, program is ready to be executed
+    - Enum variant `4u64`: `Finalized`, same as `Deployed`, but can not be
+    modified anymore
+  - `u64` Slot in which the program was last deployed, retracted or
+  initialized.
+  - `[u8; 32]` Authority address which can send program management
+  instructions.
+- Body:
+  - `[u8]` The programs executable file
+
+Verification the program account checks in the following order that:
+
+- the owner of the program account is loader-v4,
+otherwise throw `InvalidAccountOwner`
+- the program account is at least as long enough for the header,
+otherwise throw `AccountDataTooSmall`
+- the program account is writable, otherwise throw `InvalidArgument`
+- the provided authority (instruction account at index 1) signed,
+otherwise throw `MissingRequiredSignature`
+- the authority stored in the program account is the one provided,
+otherwise throw `IncorrectAuthority`
+- the status stored in the program account is not `Invalid`,
+otherwise throw `InvalidArgument`
+- the status stored in the program account is not `Finalized`,
+otherwise throw `Immutable`
```

Comment:
Should be ok to withdraw excess lamports from finalized programs right? If we wanted to allow this, we could keep the authority as-is without reseting to the system program.